sdk/identity/identity/test/public/node/caeARM.spec.ts

```diff
@@ -0,0 +1,168 @@
+  // This pipeline will trigger the CAE challenge and handle the CAE flow.
+  const pipeline = createEmptyPipeline();
+  pipeline.addPolicy(
+    bearerTokenAuthenticationPolicy({
+      scopes: [graphScope],
+      credential,
+      challengeCallbacks: {
+        async authorizeRequest({ request }): Promise<void> {
+          request.headers.set("Authorization", `Bearer ${graphToken!.token}`);
+        },
+        authorizeRequestOnChallenge: authorizeRequestOnClaimChallenge
+      }
+    })
+  );
+  const httpClient = createDefaultHttpClient();
+
+  // Sending a final request through the pipeline.
+  // On the recordings, we should see that:
+  // 1. This request fails with 401 and has a WWW-Authenticate header in its response.
+  // 2. Then we do a token request with the claims received from the challenge.
+  // 3. Then the original request is repeated and succeeds.
+  const finalResponse = await pipeline.sendRequest(
+    httpClient,
+    createPipelineRequest({
+      url: revokeUrl,
+      method: "POST"
+    })
+  );
+  assert.equal(finalResponse.status, 200, "Final response failed.");
+
+  // Getting a new token at this point should return a token different from the initially obtained.
+  // Recordings help us verify that the internal flow indeed happened.
+  const finalAccessToken = await credential.getToken(graphScope);
+  assert.ok(finalAccessToken?.token);
+  assert.ok(finalAccessToken?.expiresOnTimestamp! > Date.now());
+  resultingAccessTokens.push(finalAccessToken!);
+
+  return resultingAccessTokens;
+}
+
+describe("CAE", function() {
+  let cleanup: MsalTestCleanup;
+  beforeEach(function(this: Context) {
+    cleanup = msalNodeTestSetup(this).cleanup;
+  });
+  afterEach(async function() {
+    await cleanup();
+  });
+
+  it("DeviceCodeCredential", async function(this: Context) {
+    // Important: Recording this test may only work in certain tenants.
+
+    // Improtant:
+    // After recording these tests, the scopes in the recordings need to be changed manually.
+    // This is because MSAL is changing the scopes here: https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/a81fcef3d82523e03828d91bb0ee8d2ab2cc20d8/lib/msal-common/src/request/RequestParameterBuilder.ts#L49-L53
+    // in a way that breaks the recorder's sanitizeScopeUrl: https://github.com/Azure/azure-sdk-for-js/blob/1514d847f5a4fdb868d0407f40dcf308e5086f5a/sdk/test-utils/recorder/src/utils/index.ts#L539-L541
+
+    if (!isPlaybackMode()) {
+      this.skip();
+    }
+
+    const [firstAccessToken, finalAccessToken] = await graphChallengeFlow(
+      new DeviceCodeCredential({ tenantId: env.AZURE_TENANT_ID })
+    );
+
+    // Important:
+    // IN PLAYBACK MODE...
+    // Verifying that the first access token and the final one are different will not work consistently in this test.
+    // The recorder strips out the access tokens from the responses on the recordings.
```

Comment:
My solution in Python was to record hashes of the access tokens. That way the values are unique and no longer secret.

Reply:
Our recorder in JS strips the access token, but the service sometimes answers with different expiration dates. I could add some boilerplate code to fix the resulting access tokens to different values, but it will get really cumbersome given that the recorder doesn’t play well with other mocking tools. For unit-tests, we already have unit tests in core-client and core-rest-pipeline about the underlying code used here.